TrackletAlgorithm/TrackMerger.cc

```diff
@@ -0,0 +1,176 @@
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_2_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_3_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_0_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_1_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_2_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_3_o 
+) {
+  #pragma HLS inline
+  trackWord_o.write(aTrack._trackWord);
+  barrelStubWords_0_o.write(aTrack._barrelStubArray[0][0]);
+  barrelStubWords_1_o.write(aTrack._barrelStubArray[1][0]);
+  barrelStubWords_2_o.write(aTrack._barrelStubArray[2][0]);
+  barrelStubWords_3_o.write(aTrack._barrelStubArray[3][0]);
+  diskStubWords_0_o.write(aTrack._diskStubArray[0][0]);
+  diskStubWords_1_o.write(aTrack._diskStubArray[1][0]);
+  diskStubWords_2_o.write(aTrack._diskStubArray[2][0]);
+  diskStubWords_3_o.write(aTrack._diskStubArray[3][0]);
+
+  return;
+}
+
+void TrackMerger(const BXType bx,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::TrackWord> &trackWord,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_0,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_1,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_2,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_3,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_0,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_1,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_2,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_3,
+  BXType bx_o,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::TrackWord> &trackWord_o,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_0_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_1_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_2_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_3_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_0_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_1_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_2_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_3_o 
+  )
+{
+  #pragma HLS inline
+  static ComparisonModule comparisonModule[kNComparisonModules];
+  #pragma HLS array_partition variable=comparisonModule complete dim=0
+
+  // TrackStruct unmergedTracks[kMaxTrack];
+  // #pragma HLS array_partition variable=unmergedTracks complete dim=0
+  LOOP_Input:
+  for (unsigned int i = 0; i < kMaxTrack; i++){ 
+    #pragma HLS pipeline II=1
```

Comment:
Have you tried adding the `rewind` option here? It will be necessary eventually so that the pipeline doesn't have to drain before starting to process the next event.